### PR TITLE
FINERACT-2602: Remove orphaned COB locks after job completion

### PR DESCRIPTION
**Pull request by oleksii-novikov-onix**
## Description

Describe the changes made and why they were made. (Ignore if these details are present on the associated Apache Fineract JIRA ticket.)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per [our guidelines](https://github.com/apache/fineract/blob/develop/CONTRIBUTING.md#pull-requests)
- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.
- [ ] Create/update [unit or integration tests](https://fineract.apache.org/docs/current/#_testing) for verifying the changes made.
- [ ] Follow our [coding conventions](https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions).
- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes
- [ ] [This PR must not be a "code dump"](https://cwiki.apache.org/confluence/display/FINERACT/Pull+Request+Size+Limit). Large changes can be made in a branch, with assistance. Ask for help on the [developer mailing list](https://fineract.apache.org/#contribute).

Your assigned reviewer(s) will follow our [guidelines for code reviews](https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide).

### CODE REVIEW

File: fineract-cob/src/main/java/org/apache/fineract/cob/api/LockRequest.java [hunk 1]
```diff
@@ -18,10 +18,13 @@
...
 
+import java.time.LocalDate;
 import lombok.Data;
 
 @Data
 public class LockRequest {
 
     private String error;
+    private LocalDate cobBusinessDate;
+    private Boolean nullCobBusinessDate;
```
Comment (adamsaghy) [inline fineract-cob/src/main/java/org/apache/fineract/cob/api/LockRequest.java:29]: what is this field for?

File: fineract-cob/src/main/java/org/apache/fineract/cob/tasklet/UnlockProcessedAccountsTasklet.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+@RequiredArgsConstructor
+public abstract class UnlockProcessedAccountsTasklet<T extends AccountLock> implements Tasklet {
+
+    private final AccountLockService<T> accountLockService;
+
+    @Override
+    public RepeatStatus execute(@NonNull final StepContribution contribution, @NonNull final ChunkContext chunkContext) throws Exception {
+        final int removedCount = accountLockService.removeOrphanedLocksForProcessedAccounts();
+        if (removedCount > 0) {
+            log.info("Unlocked {} account(s) that completed COB processing but remained locked", removedCount);
```
Comment (adamsaghy) [inline fineract-cob/src/main/java/org/apache/fineract/cob/tasklet/UnlockProcessedAccountsTasklet.java]: use debug level logging please...

File: fineract-cob/src/main/java/org/apache/fineract/cob/service/AbstractAccountLockService.java [hunk 3]
```diff
@@ -59,8 +62,13 @@ public boolean isLockOverrulable(Long loanId) {
     @Transactional(propagation = Propagation.REQUIRES_NEW)
     public void updateCobAndRemoveLocks() {
         customLoanAccountLockRepository.updateLoanFromAccountLocks();
-        loanAccountLockRepository.removeByLockOwnerInAndErrorIsNotNullAndLockPlacedOnCobBusinessDateIsNotNull(
-                List.of(LockOwner.LOAN_COB_CHUNK_PROCESSING, LockOwner.LOAN_INLINE_COB_PROCESSING));
+        loanAccountLockRepository.removeByLockOwnerInAndErrorIsNotNullAndLockPlacedOnCobBusinessDateIsNotNull(COB_LOCK_OWNERS);
+    }
+
+    @Override
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
```
Comment (adamsaghy) [inline fineract-cob/src/main/java/org/apache/fineract/cob/service/AbstractAccountLockService.java]: Do we need to start a new transaction for this?

**Comment by AshharAhmadKhan** [comment]
@oleksii-novikov-onix @adamsaghy 

The approach here is solid. A few thoughts on the open review comments:

`@Transactional(propagation = REQUIRES_NEW)` on `removeOrphanedLocksForProcessedAccounts()` isn't needed. Spring Batch already wraps each tasklet in its own transaction and all prior steps have committed by the time this runs, so default `REQUIRED` is fine.

`nullCobBusinessDate` in `LockRequest` is there so tests can place a lock with a null cob date through the API, which is needed to cover the `lock_placed_on_cob_business_date IS NOT NULL` filter in the delete query. Without it there's no way to test that edge case.

Also agree on `log.info`, `log.debug` makes more sense since this runs on every COB execution.

File: fineract-cob/src/main/java/org/apache/fineract/cob/tasklet/UnlockProcessedAccountsTasklet.java (hunk 2 above)
Comment (oleksii-novikov-onix) [reply fineract-cob/src/main/java/org/apache/fineract/cob/tasklet/UnlockProcessedAccountsTasklet.java]: Fixed

File: fineract-cob/src/main/java/org/apache/fineract/cob/service/AbstractAccountLockService.java (hunk 3 above)
Comment (oleksii-novikov-onix) [reply fineract-cob/src/main/java/org/apache/fineract/cob/service/AbstractAccountLockService.java]: Right, we do not need REQUIRES_NEW here. This method is invoked only from UnlockProcessedAccountsTasklet.execute().

Comment (oleksii-novikov-onix) [reply fineract-cob/src/main/java/org/apache/fineract/cob/service/AbstractAccountLockService.java]: Removed

**Comment by oleksii-novikov-onix** [comment]
@AshharAhmadKhan Thank you for participating in the discussion

File: fineract-cob/src/main/java/org/apache/fineract/cob/api/LockRequest.java (hunk 1 above)
Comment (oleksii-novikov-onix) [reply fineract-cob/src/main/java/org/apache/fineract/cob/api/LockRequest.java:29]: LockRequest was added only for the internal API, specifically for testing locks. I decided to extend it to also simulate "stale" lock records.

Comment (oleksii-novikov-onix) [reply fineract-cob/src/main/java/org/apache/fineract/cob/api/LockRequest.java:29]: JSON does not allow distinguishing three states:
1. {} - field is not provided, so COB date should be taken from ThreadLocal as default
2. {"cobBusinessDate": "2026-05-20"} - an explicit date is provided
3. {"cobBusinessDate": null} - explicitly store NULL in the database

After deserialization, states 1 and 3 look identical

**Review by adamsaghy — APPROVED** [review]
LGTM
